### PR TITLE
(BKR-203) Add helpers to create custom facts

### PR DESCRIPTION
**Pull request by electrical**
Opening this for early feedback.
This allows to create custom facts and utilize those with hiera for example.

### CODE REVIEW

**Comment by electrical** [comment]
Just a note, it only works on linux/unix.
I wasn't sure what directories to use for pswindows and windows hosts ( and others )
Would love to add those as well before merging.

**Comment by anodelman** [comment]
@joshcooper Do you know the windows path for the facter dir?

**Comment by liamjbennett** [comment]
https://docs.puppetlabs.com/facter/2.2/custom_facts.html#fact-locations

**Comment by electrical** [comment]
@liamjbennett thanks!
Since we don't do much specifics for windows versions, any thoughts on doing that? Or just ignore the win2003 path?

**Comment by liamjbennett** [comment]
My suggestion would be to ignore to win2003 path for now. If anyone still cares about win2003 support (I might do one day - but most people won't) then we can add it then.

**Comment by electrical** [comment]
Cool. Added the paths. Could you take a peek if its okay like this?

**Comment by liamjbennett** [comment]
:+1:

**Comment by anodelman** [comment]
ok to test

**Comment by anodelman** [comment]
Syntax errors during yard doc generation:
```
[warn]: Syntax error in `lib/beaker/host/windows.rb`:(32,9): syntax error, unexpected tstring_beg, expecting '}'
[warn]: @param tag has unknown parameter name: The 
    in file `lib/beaker/dsl/helpers/facter_helpers.rb' near line 52
```

**Comment by electrical** [comment]
@anodelman fixed it. i was being stupid and forgot 2 comma's to place.

**Comment by anodelman** [comment]
Syntax errors during yard doc generation:

```
[warn]: @param tag has unknown parameter name: The 
    in file `lib/beaker/dsl/helpers/facter_helpers.rb' near line 52
```

**Comment by electrical** [comment]
I really should read the error and fix the actual thing :p

**Comment by electrical** [comment]
Yay :)

File: lib/beaker/host/windows.rb [hunk 1]
```diff
@@ -46,6 +47,7 @@ def self.foss_defaults
         'puppetvardir'      => '`cygpath -smF 35`/PuppetLabs/puppet/var',
         'distmoduledir'     => '`cygpath -smF 35`/PuppetLabs/puppet/etc/modules',
         'sitemoduledir'     => 'C:/usr/share/puppet/modules',
+        'facter-factsd'     => '/cygdrive/c/ProgramData/PuppetLabs/facter/facts.d/',
```
Comment (anodelman) [inline lib/beaker/host/windows.rb]: This should use cygpath to work on both 64 and 32.

**Comment by anodelman** [comment]
@kevpl can you look into this to see if there is any other way of determining the facter dir at runtime?

File: lib/beaker/host/windows.rb (hunk 1 above)
Comment (electrical) [inline lib/beaker/host/windows.rb]: should that be something like ``'`cygpath -smF 35`/PuppetLabs/facter/facts.d/' `` ?

**Comment by anodelman** [comment]
Last review comments still current.

**Comment by kevpl** [comment]
the only way that I'm seeing to do this is to use a file on the system, but I'm not very familiar with it.

@joshcooper, do you know of a way to add custom facts that doesn't involve adding another path to Beaker?  We could just have the user need to pass in the facts.d path like with_puppet_running_on & puppetserver-confdir. Making that method more general could help us if there are custom facter paths, but I don't know if people do that...

@electrical my apologies for the delay here.  In our latest work, we've been trying to eliminate the need for these paths in the platform ruby files, as they can block development work if they need changing. That's why the delay.  In the meantime, would you mind cleaning up the commit messages from Apr 17th?  Hopefully we'll be able to finish this conversation soon, and get these changes in.

**Comment by electrical** [comment]
@kevpl no worries. I've squashed the commits into a single one.

**Comment by kevpl** [comment]
failure unrelated to changes.

@electrical thanks for understanding.  If we don't have a good recommendation on a way to do this without the paths by Monday, I'll merge this as is.

**Comment by electrical** [comment]
@kevpl Cool thanks!

**Comment by joshcooper** [comment]
There are two external facts directories, one that facter always looks in, e.g. `/etc/facter/facts.d`, and one managed by puppet, e.g. `/var/lib/puppet/lib/facts.d`. Puppet will pluginsync external facts from modules into the latter directory, and add that search path to facter, so facter is able to resolve pluginsynced external facts.

The latter directory can be obtained via `puppet agent --configprint pluginfactdest` though that won't help this PR.

This PR adds helpers for the former, and it's location has been hardcoded in facter (both ruby and native implementations) since at least 1.7.0: https://github.com/puppetlabs/facter/blob/1.7.0/lib/facter/util/config.rb#L35-L37. Note facter's search path includes both FOSS and PE locations. 

I don't know of a way for beaker to query for these locations, but since they have been fixed, I don't see harm in beaker hardcoding them. The only problem I see with this PR, is it doesn't account for AIO paths.

**Comment by kevpl** [comment]
@joshcooper, thanks for your help!

@electrical, you should be able to account for Josh's concerns by adding the `facter-factsd` path to the `aio_defaults` as well.  Once that's done, we'll get it through final testing, and merging.

**Comment by electrical** [comment]
@kevpl will have that fixed before Monday so it can be reviewed further.

Cheers.

**Comment by anodelman** [comment]
@electrical can you look over freebsd.rb, pswindows.rb, mac.rb, etc to get their pathing updated to include the facter stuff?

**Comment by electrical** [comment]
@anodelman i would love to but its hard to find what the right paths would need to be :-)
I tried to dig into the facter code but I'm nothing wiser yet.

@joshcooper @kevpl Do you guys have any resources i can look at for finding the right paths?
https://docs.puppetlabs.com/facter/2.4/custom_facts.html#fact-locations has a few ( like the mac ) but nothing for freebsd so not sure what to select there.

Thanks!

File: lib/beaker/dsl/helpers/facter_helpers.rb [hunk 2]
```diff
@@ -42,6 +42,24 @@ def fact(name, opts = {})
...
+        # Create a custom fact on the provided host
+        # We do this by setting a static value in the facts.d directory
+        #
+        # @param [Host, Array<Host>, String, Symbol] host    One or more hosts to act upon,
+        #                            or a role (String or Symbol) that identifies one or more hosts.
+        # @param [String] name The name of the fact to create
+        # @param [String, Array] value  The value of the fact to create
+        def create_fact_on(host, name, value)
+          block_on host do | host |
+            create_remote_file(host, File.join(host.puppet['facter-factsd'], "#{name}.yaml"), value.to_yaml)
```
Comment (kevpl) [inline lib/beaker/dsl/helpers/facter_helpers.rb]: @electrical, looking through this, I'm seeing that you're using `host.puppet['facter-factsd']` in order to get the paths.  That call won't work for paths defined in the host defaults, that tries to get the values from a puppet configprint call.  In order to get the values you've defined, you'll have to change that to `host['facter-factsd']`.

Comment (electrical) [inline lib/beaker/dsl/helpers/facter_helpers.rb]: Ah okay, will fix that :-)

**Comment by anodelman** [comment]
Sorry, @electrical.  We did a big merge and looks like you need a rebase now.

**Comment by anodelman** [comment]
Last comment still current.

**Comment by anodelman** [comment]
Will close out next week due to inactivity.

**Comment by electrical** [comment]
Sorry work has been very busy. Will make some time for it this week.

**Comment by electrical** [comment]
Rebase went completely wrong. Fixing :-)

**Comment by thomasmeeus** [comment]
Hi @electrical is this still work in progress? I would love to see this merged. Thx

**Comment by electrical** [comment]
Hi,

Completely forgot about this.
Will pick this up again in the weekend and send a new PR.

**Comment by thomasmeeus** [comment]
:+1:

**Pull request closed without merging**